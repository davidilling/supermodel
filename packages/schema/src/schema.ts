import { JSONSchema7 } from 'json-schema';
import { SchemaFileReader } from '@supermodel/fs';
import { schemaFetch } from '@supermodel/http';
import { validateSchema } from '@supermodel/validator';
import { SchemaSource, Url } from './schema/utils';
import { ResolverOptions, SchemaResolver } from './schema/resolver';

export class Schema {
  // TODO: circular detection
  circular: boolean = false;

  entrySchema?: JSONSchema7;
  resolvedSchemas?: JSONSchema7[] = [];

  private source: SchemaSource;
  private options: ResolverOptions;

  constructor(source: SchemaSource, options: ResolverOptions = {}) {
    this.source = source;
    this.options = options;
  }

  get resolved() {
    return this.resolvedSchemas !== undefined;
  }

  async resolve() {
    const resolver = new SchemaResolver(this.source, this.options);
    Object.assign(this, await resolver.resolve());
  }

  async bundle() {
    if (!this.resolved) {
      await this.resolve();
    }
  }
}
