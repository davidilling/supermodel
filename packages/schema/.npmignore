/src
/test
/fixtures
/scripts
/tsconfig.json
/.env
/.env.development