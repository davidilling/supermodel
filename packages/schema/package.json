{
  "name": "@supermodel/schema",
  "version": "0.0.1",
  "description": "Schema core for Supermodel",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/supermodel/supermodel.git"
  },
  "author": {
    "name": "Zdenek Nemec",
    "email": "z@goodapi.co",
    "url": "http://goodapi.co"
  },
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/supermodel/supermodel/issues"
  },
  "homepage": "https://github.com/supermodel/supermodel/tree/master/packages/file#readme",
  "dependencies": {
    "@supermodel/fs": "0.0.1",
    "@supermodel/http": "0.0.1",
    "@supermodel/validator": "^0.0.1",
    "json-schema-traverse": "^0.4.1",
    "path": "0.12.7",
    "resolve-pathname": "^3.0.0",
    "url": "0.11.0"
  },
  "files": [
    "lib"
  ],
  "devDependencies": {
    "@types/json-schema": "^7.0.3",
    "fast-clone": "^1.5.13",
    "typescript": "^3.3.3"
  },
  "optionalDependencies": {
    "@supermodel/file": "^0.5.3"
  }
}
