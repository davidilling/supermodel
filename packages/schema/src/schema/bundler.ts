import clone = require('fast-clone');
import { JSONSchema7 } from 'json-schema';
import traverse = require('json-schema-traverse');

export class SchemaBundler {
  schema: JSONSchema7;
  schemas: { [schemaId: string]: JSONSchema7 };

  constructor(
    entrySchema: JSONSchema7,
    schemas: { [schemaId: string]: JSONSchema7 },
  ) {
    this.schema = clone(entrySchema);
    this.schemas = clone(schemas);
  }

  bundle() {
    const definitions = this.schema.definitions || {};

    for (const schemaId in this.schemas) {
      if (this.schemas.hasOwnProperty(schemaId)) {
        const schema = this.schemas[schemaId];

        if (schema.definitions) {
          for (const definitionKey in schema.definitions) {
            if (schema.definitions.hasOwnProperty(definitionKey)) {
              definitions[definitionKey] = schema.definitions[definitionKey];
            }
          }
        }
      }
    }

    this.schema.definitions = definitions;

    return this.schema;
  }
}
