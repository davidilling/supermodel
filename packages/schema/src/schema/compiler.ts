import { JSONSchema7 } from 'json-schema';

export class SchemaCompiler {
  constructor(entrySchema: JSONSchema7, schemas: JSONSchema7[]) {}

  bundle() {}
}
