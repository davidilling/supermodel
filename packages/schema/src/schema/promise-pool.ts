export type Worker = (data: any) => Promise<any>;
export type DataFetcher = () => any | undefined;

export class PromisePool {
  private worker: Worker;
  private dataFetcher: DataFetcher;
  private concurrency: number;
  private poolSize: number;
  private callbacks?: { reject: (error: Error) => void; resolve: () => void };
  private done: boolean;

  constructor(worker: Worker, dataFetcher: DataFetcher, concurrency: number) {
    this.worker = worker;
    this.dataFetcher = dataFetcher;
    this.concurrency = concurrency;
    this.poolSize = 0;
    this.done = false;
  }

  start() {
    return new Promise((resolve, reject) => {
      this.callbacks = { reject, resolve };
      this.startPool();
    });
  }

  abort(err = new Error('Abort pool without particular reason')) {
    if (!this.callbacks) {
      throw new Error("Can't abort unstarted pool");
    }

    if (this.done) {
      throw new Error("Can't abort finished pool");
    }

    this.done = true;
    this.callbacks.reject(err);
  }

  private async fetchData() {
    return await this.dataFetcher();
  }

  private startPool() {
    for (let i = 0; i < this.concurrency; i++) {
      this.buildWorker(Promise.resolve());
    }
  }

  private buildWorker(promise: Promise<any>) {
    ++this.poolSize;

    promise.then(this.onWorkerDone).catch(this.onWorkerFail);
  }

  private onWorkerDone = async () => {
    --this.poolSize;

    if (!this.done) {
      const data = await this.fetchData();

      if (data) {
        this.buildWorker(this.worker(data));
      } else if (this.poolSize === 0 && this.callbacks) {
        this.done = true;
        this.callbacks.resolve();
      }
    }
  };

  private onWorkerFail = (error: Error) => {
    if (this.callbacks) {
      this.done = true;
      // TODO: wait for pending workers and then reject
      this.callbacks.reject(error);
    }
  };
}
