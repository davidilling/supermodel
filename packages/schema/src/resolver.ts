import { JSONSchema7 } from 'json-schema';
import { SchemaFileReader } from '@supermodel/fs';
import { schemaFetch } from '@supermodel/http';
import { validateSchema } from '@supermodel/validator';
import {
  SchemaSource,
  isUrl,
  Url,
  collectRefs,
  collectDefinitions,
} from './utils';
import { PromisePool } from './promise-pool';

export type ResolverOptions = {
  cwd?: string;
  file?: typeof SchemaFileReader;
  http?: boolean;
  concurrency?: number;
  validate?: boolean;
  circular?: boolean;
  schemaId?: boolean;
};

type InternalResolverOptions = {
  http: boolean;
  concurrency: number;
  validate: boolean;
  circular: boolean;
  schemaId: boolean;
};

type SchemasCache = Map<Url, Promise<JSONSchema7>>;

type Queue = Array<string>;

export class SchemaResolver {
  resolvedSchemas: JSONSchema7[] = [];

  // TODO: circular detection
  circular: boolean = false;
  resolved: boolean = false;

  private initialSchema?: JSONSchema7;
  private options: ResolverOptions & InternalResolverOptions;
  private local: boolean = false;
  private schemaFileInstance?: SchemaFileReader;
  private source: SchemaSource;
  private pool?: PromisePool | null;

  constructor(source: SchemaSource, options: ResolverOptions = {}) {
    this.options = {
      http: true,
      concurrency: 10,
      validate: true,
      circular: true,
      schemaId: false,
      ...options,
    };

    this.source = source;
  }

  async resolve() {
    this.initialSchema = await this.getInitialSchema();
    const pendingSchemas: SchemasCache = new Map();
    const queue: Queue = [];

    this.resolveSchema(
      this.initialSchema.$id,
      this.initialSchema,
      pendingSchemas,
      queue,
    );

    await this.processQueue(queue, pendingSchemas, this.options.concurrency);

    this.resolvedSchemas = await Promise.all(pendingSchemas.values());

    return (this.resolved = true);
  }

  async bundle() {
    if (!this.resolved) {
      await this.resolve();
    }
  }

  /**
   * Get first schema from new SchemaResolver(schema). Could be object, url or local path
   */
  private async getInitialSchema() {
    if (typeof this.source === 'object') {
      return this.source;
    } else if (typeof this.source === 'string') {
      if (isUrl(this.source)) {
        return await schemaFetch(this.source);
      }

      this.local = true;
      return await this.file.read(this.source);
    }

    throw new Error(`Can\'t process source of type '${typeof this.source}'.`);
  }

  /**
   * Get others schemas. If you provide options.file = SchemaFile and first schema
   * was local it try to load locally first and when missing it fetch from http.
   * You disable http fetch at all with options.http = false
   */
  private async getSchema(id: string) {
    if (this.local || !this.options.http) {
      const schema = await this.file.readId(id, !this.options.http);

      if (schema) {
        return schema;
      }
    }

    return await schemaFetch(id);
  }

  private resolveSchema(
    schemaId: string | undefined,
    schema: JSONSchema7,
    pendingSchemas: SchemasCache,
    queue: Queue,
  ) {
    if (this.options.validate) {
      try {
        validateSchema(schema);
      } catch (err) {
        if (this.pool) {
          this.pool.abort(err);
        }

        throw err;
      }
    }

    this.addSchemaToCache(schemaId, schema, pendingSchemas);

    // Collect definition from model to resolved cache
    collectDefinitions(schema).forEach(definition =>
      this.addSchemaToCache(definition.$id, definition, pendingSchemas),
    );

    // Collect refs and enque them for resolve
    collectRefs(schema).forEach(id =>
      this.enqueueSchema(id, pendingSchemas, queue),
    );
  }

  /**
   * Put schema id into queue to get
   */
  private async enqueueSchema(
    schemaId: string,
    pendingSchemas: SchemasCache,
    queue: Queue,
  ) {
    if (!pendingSchemas.has(schemaId) && !queue.includes(schemaId)) {
      queue.push(schemaId);
    }
  }

  private async processQueue(
    queue: Queue,
    pendingSchemas: SchemasCache,
    concurrency: number,
  ) {
    this.pool = new PromisePool(
      this.makeQueueWorker(pendingSchemas, queue),
      this.makeQueueDataFetcher(queue),
      concurrency,
    );

    return this.pool.start();
  }

  /**
   * Makes function for pool which get result from data fetcher and process it
   */
  private makeQueueWorker(pendingSchemas: SchemasCache, queue: Queue) {
    return async (schemaId: string) => {
      const schema = await this.getSchema(schemaId);
      this.resolveSchema(schemaId, schema, pendingSchemas, queue);
    };
  }

  /**
   * Makes function for pool which picks schemas from queue to resolve
   */
  private makeQueueDataFetcher(queue: Queue) {
    return () => queue.shift();
  }

  private addSchemaToCache(
    schemaId: string | undefined,
    schema: JSONSchema7,
    pendingSchemas: SchemasCache,
  ) {
    const $id = schema.$id || schemaId;

    if ($id) {
      if (!pendingSchemas.has($id)) {
        pendingSchemas.set($id, Promise.resolve(schema));
      } else {
        // TODO: we could check that content of schema is same as previous one. otherwise throw error
      }
    } else if (this.options.schemaId) {
      throw new Error('Resolved schema is missing $id');
    }
  }

  // File handling
  private get file() {
    if (!this.schemaFileInstance) {
      const { file: SchemaFileClass } = this.options;

      if (!SchemaFileClass) {
        throw new Error(`
          To use local resolving, you must provide 'file' option to Resolver constructor
          'new Resolver(path, { file: require("@supermodel/file")})' or with custom one.
        `);
      }

      this.schemaFileInstance = new SchemaFileClass(this.options.cwd);
    }

    if (!this.schemaFileInstance) {
      throw new Error('Something went terribly wrong');
    }

    return this.schemaFileInstance;
  }
}
