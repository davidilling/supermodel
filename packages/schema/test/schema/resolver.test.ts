import { SchemaFileReader } from '@supermodel/fs';
import { SchemaResolver } from '../../src/schema/resolver';
import { resolve } from 'path';
import { JSONSchema7 } from 'json-schema';

require('jest-playback').setup(process.cwd(), 'run');

const sortSchemas = (schemas: JSONSchema7[]) => {
  return schemas.sort((a, b) => {
    if ((a.$id as string) < (b.$id as string)) {
      return -1;
    }
    if ((a.$id as string) > (b.$id as string)) {
      return 1;
    }
    return 0;
  });
};

describe('SchemaResolver', () => {
  // NOTE: Resolving whole schemaorg is suicide in tests...
  // But can be used for testing :)

  // test('resolve valid schema via http', async () => {
  //   const resolver = new Schema(
  //     'https://supermodel.io/schemaorg/Action',
  //     {
  //       concurrency: 5,
  //     },
  //   );
  //   const result = await resolver.resolve();
  //   expect(sortSchemas(result)).toMatchSnapshot();
  // }, 100000);

  test('resolve valid schema via http #2', async () => {
    const resolver = new SchemaResolver(
      'https://supermodel.io/adidas/product/Article',
    );
    const { schemas } = await resolver.resolve();
    expect(sortSchemas(Object.values(schemas))).toMatchSnapshot();
  });

  test('resolve valid schema layer via http #3', async () => {
    const resolver = new SchemaResolver('https://supermodel.io/adidas/product');
    const { schemas } = await resolver.resolve();
    expect(sortSchemas(Object.values(schemas))).toMatchSnapshot();
  });

  test('resolve valid schema via file', async () => {
    const ActionSchemaPath = resolve(
      __dirname,
      '../../../../fixtures/schema/schemaorg/Action.yaml',
    );
    const resolver = new SchemaResolver(ActionSchemaPath, {
      file: SchemaFileReader,
    });
    const { schemas } = await resolver.resolve();
    expect(sortSchemas(Object.values(schemas))).toMatchSnapshot();
  });

  test('resolve valid schema via directory', async () => {
    const ActionSchemaPath = resolve(
      __dirname,
      '../../../../fixtures/schema/schemaorg',
    );
    const resolver = new SchemaResolver(ActionSchemaPath, {
      file: SchemaFileReader,
    });
    const { schemas } = await resolver.resolve();
    expect(sortSchemas(Object.values(schemas))).toMatchSnapshot();
  });
});
