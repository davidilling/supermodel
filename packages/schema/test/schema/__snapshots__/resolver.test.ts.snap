// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaResolver resolve valid schema layer via http #3 1`] = `Array []`;

exports[`SchemaResolver resolve valid schema via directory 1`] = `
Array [
  Object {
    "$id": "http://supermodel.io/schemaorg/Action",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/Action",
    "allOf": Array [
      Object {
        "$ref": "http://supermodel.io/schemaorg/Thing",
      },
    ],
    "description": "An action performed by a direct agent and indirect participants upon a direct object. Optionally happens at a location with the help of an inanimate instrument. The execution of the action may produce a result. Specific action sub-type documentation specifies the exact expectation of each argument/role.<br/><br/>

See also <a href=\\"http://blog.schema.org/2014/04/announcing-schemaorg-actions.html\\">blog post</a> and <a href=\\"http://schema.org/docs/actions.html\\">Actions overview document</a>.",
    "properties": Object {
      "error": Object {
        "$ref": "http://supermodel.io/schemaorg/properties/error",
      },
      "result": Object {
        "$ref": "http://supermodel.io/schemaorg/properties/result",
      },
    },
    "title": "Action",
    "type": "object",
  },
  Object {
    "$id": "http://supermodel.io/schemaorg/DataType",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/DataType",
    "description": "The basic data types such as Integers, Strings, etc.",
    "title": "DataType",
  },
  Object {
    "$id": "http://supermodel.io/schemaorg/Text",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/Text",
    "allOf": Array [
      Object {
        "$ref": "http://supermodel.io/schemaorg/DataType",
      },
    ],
    "description": "Data type: Text.",
    "title": "Text",
    "type": "string",
  },
  Object {
    "$id": "http://supermodel.io/schemaorg/Thing",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/Thing",
    "description": "The most generic type of item.",
    "properties": Object {
      "name": Object {
        "$ref": "http://supermodel.io/schemaorg/properties/name",
      },
    },
    "title": "Thing",
    "type": "object",
  },
  Object {
    "$id": "http://supermodel.io/schemaorg/properties/error",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/error",
    "description": "For failed actions, more information on the cause of the failure.",
    "oneOf": Array [
      Object {
        "$ref": "http://supermodel.io/schemaorg/Thing",
      },
    ],
    "title": "error",
  },
  Object {
    "$id": "http://supermodel.io/schemaorg/properties/name",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/name",
    "description": "The name of the item.",
    "oneOf": Array [
      Object {
        "$ref": "http://supermodel.io/schemaorg/Text",
      },
    ],
    "title": "name",
  },
  Object {
    "$id": "http://supermodel.io/schemaorg/properties/result",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/result",
    "description": "The result produced in the action. e.g. John wrote <em>a book</em>.",
    "oneOf": Array [
      Object {
        "$ref": "http://supermodel.io/schemaorg/Thing",
      },
    ],
    "title": "result",
  },
]
`;

exports[`SchemaResolver resolve valid schema via file 1`] = `
Array [
  Object {
    "$id": "http://supermodel.io/schemaorg/Action",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/Action",
    "allOf": Array [
      Object {
        "$ref": "http://supermodel.io/schemaorg/Thing",
      },
    ],
    "description": "An action performed by a direct agent and indirect participants upon a direct object. Optionally happens at a location with the help of an inanimate instrument. The execution of the action may produce a result. Specific action sub-type documentation specifies the exact expectation of each argument/role.<br/><br/>

See also <a href=\\"http://blog.schema.org/2014/04/announcing-schemaorg-actions.html\\">blog post</a> and <a href=\\"http://schema.org/docs/actions.html\\">Actions overview document</a>.",
    "properties": Object {
      "error": Object {
        "$ref": "http://supermodel.io/schemaorg/properties/error",
      },
      "result": Object {
        "$ref": "http://supermodel.io/schemaorg/properties/result",
      },
    },
    "title": "Action",
    "type": "object",
  },
  Object {
    "$id": "http://supermodel.io/schemaorg/DataType",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/DataType",
    "description": "The basic data types such as Integers, Strings, etc.",
    "title": "DataType",
  },
  Object {
    "$id": "http://supermodel.io/schemaorg/Text",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/Text",
    "allOf": Array [
      Object {
        "$ref": "http://supermodel.io/schemaorg/DataType",
      },
    ],
    "description": "Data type: Text.",
    "title": "Text",
    "type": "string",
  },
  Object {
    "$id": "http://supermodel.io/schemaorg/Thing",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/Thing",
    "description": "The most generic type of item.",
    "properties": Object {
      "name": Object {
        "$ref": "http://supermodel.io/schemaorg/properties/name",
      },
    },
    "title": "Thing",
    "type": "object",
  },
  Object {
    "$id": "http://supermodel.io/schemaorg/properties/error",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/error",
    "description": "For failed actions, more information on the cause of the failure.",
    "oneOf": Array [
      Object {
        "$ref": "http://supermodel.io/schemaorg/Thing",
      },
    ],
    "title": "error",
  },
  Object {
    "$id": "http://supermodel.io/schemaorg/properties/name",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/name",
    "description": "The name of the item.",
    "oneOf": Array [
      Object {
        "$ref": "http://supermodel.io/schemaorg/Text",
      },
    ],
    "title": "name",
  },
  Object {
    "$id": "http://supermodel.io/schemaorg/properties/result",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$source": "http://schema.org/result",
    "description": "The result produced in the action. e.g. John wrote <em>a book</em>.",
    "oneOf": Array [
      Object {
        "$ref": "http://supermodel.io/schemaorg/Thing",
      },
    ],
    "title": "result",
  },
]
`;

exports[`SchemaResolver resolve valid schema via http #2 1`] = `
Array [
  Object {
    "$id": "http://supermodel.io/adidas/product/Article",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "examples": Array [
      Object {
        "ageGroup": Object {
          "code": "A",
          "description": "ADULT",
        },
        "articleDescription": "COPA MUNDIAL",
        "articleNumber": "015110",
        "assets": Array [
          Object {},
        ],
        "brand": Object {
          "code": "11",
          "description": "ADIDAS",
        },
        "catalog": Object {},
        "color1": Object {
          "code": "R1",
          "description": "Red1",
        },
        "color2": Object {
          "code": "R2",
          "description": "Red2",
        },
        "color3": Object {
          "code": "R3",
          "description": "Red3",
        },
        "color4": Object {
          "code": "R4",
          "description": "Red4",
        },
        "colors": Array [
          Object {},
        ],
        "colorwayDescription": "BLABAS/GRAFIT/BLABAS",
        "colorwayname": "FTWR WHITE/BLACK",
        "gender": Object {
          "code": "M",
          "description": "MALE",
        },
        "imageLaunchDate": "launchDate",
        "materialWeight": "weight",
        "modelNumber": "10034",
        "productType": Object {
          "code": "104",
          "description": "FOOTBALL SHOES (FIRM GROUND)",
        },
        "seasons": Array [
          Object {},
        ],
        "skus": Array [
          Object {},
        ],
      },
    ],
    "properties": Object {
      "ageGroup": Object {
        "$ref": "utils/CodeDescription",
      },
      "articleDescription": Object {
        "type": "string",
      },
      "articleNumber": Object {
        "type": "string",
      },
      "assets": Object {
        "items": Object {
          "$ref": "Asset",
        },
        "type": "array",
      },
      "brand": Object {
        "$ref": "utils/CodeDescription",
      },
      "businessSegment": Object {
        "$ref": "utils/CodeDescription",
      },
      "catalog": Object {
        "$ref": "Catalog",
      },
      "color1": Object {
        "$ref": "utils/CodeDescription",
      },
      "color2": Object {
        "$ref": "utils/CodeDescription",
      },
      "color3": Object {
        "$ref": "utils/CodeDescription",
      },
      "color4": Object {
        "$ref": "utils/CodeDescription",
      },
      "colors": Object {
        "items": Object {
          "$ref": "Color",
        },
        "type": "array",
      },
      "colorwayDescription": Object {
        "type": "string",
      },
      "colorwayName": Object {
        "type": "string",
      },
      "franchise": Object {
        "$ref": "utils/CodeDescription",
      },
      "gender": Object {
        "$ref": "utils/CodeDescription",
      },
      "imageLaunchDate": Object {
        "format": "date-time",
        "type": "string",
      },
      "lifecycleStatus": Object {
        "properties": Object {
          "code": Object {
            "type": "string",
          },
          "date": Object {
            "format": "date-time",
            "type": "string",
          },
          "description": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "materialWeight": Object {
        "type": "string",
      },
      "modelNumber": Object {
        "type": "string",
      },
      "productDivision": Object {
        "$ref": "utils/CodeDescription",
      },
      "productFit": Object {
        "$ref": "utils/CodeDescription",
      },
      "productGroup": Object {
        "$ref": "utils/CodeDescription",
      },
      "productType": Object {
        "$ref": "utils/CodeDescription",
      },
      "related": Object {
        "items": Object {
          "$ref": "Article",
        },
        "type": "array",
      },
      "seasons": Object {
        "items": Object {
          "$ref": "ArticleSeason",
        },
        "type": "array",
      },
      "skus": Object {
        "items": Object {
          "$ref": "SKU",
        },
        "type": "array",
      },
      "sportsCategory": Object {
        "$ref": "utils/CodeDescription",
      },
      "subBrand": Object {
        "$ref": "utils/CodeDescription",
      },
    },
    "title": "Article",
    "type": "object",
  },
  Object {
    "$id": "http://supermodel.io/adidas/product/ArticleSeason",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "description": "Seasonal information about article",
    "examples": Array [
      Object {
        "carryOver": false,
        "retailExitDate": "2017-11-30",
        "retailIntroDate": "2017-06-01",
        "seasonName": "FW17",
      },
    ],
    "properties": Object {
      "carryOver": Object {
        "type": "boolean",
      },
      "retailExitDate": Object {
        "format": "date-time",
        "type": "string",
      },
      "retailIntroDate": Object {
        "format": "date-time",
        "type": "string",
      },
      "seasonName": Object {
        "type": "string",
      },
    },
    "title": "Article Season",
    "type": "object",
  },
  Object {
    "$id": "http://supermodel.io/adidas/product/Asset",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "description": "Digital Asset",
    "examples": Array [
      Object {
        "assetCategory": "Photography",
        "assetHref": "http://domain.com/image.jpeg",
        "assetType": "image",
        "defaultAlias": "alias",
        "image": Object {
          "bgColor": "transparent",
          "height": 2000,
          "representation": null,
          "width": 2000,
        },
        "imageStyle": "Standard",
        "model": Object {
          "chest": 97,
          "height": 197,
          "productSize": 540,
          "waist": 58,
        },
        "pairedArticles": Object {
          "value": "AB1234",
        },
        "publicID": "AZ1",
        "sortOrder": 6,
        "usage": Object {
          "value": "pdp",
        },
        "usageTerms": Object {
          "value": "B2B Catalogue",
        },
        "video": Object {
          "fileSize": 99,
        },
        "view": "Top",
        "viewGroup": "3d",
      },
    ],
    "properties": Object {
      "assetCategory": Object {
        "type": "string",
      },
      "assetHref": Object {
        "format": "uri",
        "type": "string",
      },
      "assetType": Object {
        "enum": Array [
          "image",
          "video",
          "other",
        ],
        "type": "string",
      },
      "defaultAlias": Object {
        "type": "string",
      },
      "image": Object {
        "properties": Object {
          "bgColor": Object {
            "type": "string",
          },
          "height": Object {
            "type": "integer",
          },
          "representation": Object {
            "type": "string",
          },
          "width": Object {
            "type": "integer",
          },
        },
        "type": "object",
      },
      "imageStyle": Object {
        "type": "string",
      },
      "model": Object {
        "properties": Object {
          "chest": Object {
            "type": "string",
          },
          "height": Object {
            "type": "string",
          },
          "productSize": Object {
            "type": "string",
          },
          "waist": Object {
            "type": "string",
          },
        },
        "type": "object",
      },
      "pairedArticles": Object {
        "items": Array [
          Object {
            "type": "string",
            "value": null,
          },
        ],
        "type": "array",
      },
      "publicId": Object {
        "type": "string",
      },
      "sortOrder": Object {
        "type": "string",
      },
      "usage": Object {
        "items": Array [
          Object {
            "type": "string",
            "value": null,
          },
        ],
        "type": "array",
      },
      "usageTerms": Object {
        "items": Array [
          Object {
            "type": "string",
            "value": null,
          },
        ],
        "type": "array",
      },
      "video": Object {
        "properties": Object {
          "fileSize": Object {
            "type": "integer",
          },
        },
        "type": "object",
      },
      "view": Object {
        "type": "string",
      },
      "viewGroup": Object {
        "type": "string",
      },
    },
    "required": Array [
      "assetType",
      "viewGroup",
      "view",
      "assetCategory",
      "usageTerms",
      "assetHref",
    ],
    "title": "Asset",
    "type": "object",
  },
  Object {
    "$id": "http://supermodel.io/adidas/product/Catalog",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "definitions": Object {
      "CatalogCopy": Object {
        "properties": Object {
          "articleNumber": Object {
            "type": "string",
          },
          "catalog": Object {
            "items": Object {
              "$ref": "Catalogues",
            },
            "type": "array",
          },
          "modelNumber": Object {
            "type": "string",
          },
        },
        "title": "Catalog Copy",
        "type": "object",
      },
    },
    "examples": Array [
      Object {
        "catalogueCopy": Object {},
        "featureAndBenefits": Array [
          Object {},
        ],
      },
    ],
    "properties": Object {
      "catalogueCopy": Object {
        "$ref": "#/definitions/CatalogCopy",
      },
    },
    "title": "Catalog",
    "type": "object",
  },
  Object {
    "$id": "http://supermodel.io/adidas/product/Catalogues",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "properties": Object {
      "acsModelName": Object {
        "type": "string",
      },
      "b2bCopy": Object {
        "type": "string",
      },
      "b2bCopyShort": Object {
        "type": "string",
      },
      "b2cCopy": Object {
        "type": "string",
      },
      "b2cCopyShort": Object {
        "type": "string",
      },
      "b2cModelName": Object {
        "type": "string",
      },
      "bulletFit": Object {
        "type": "string",
      },
      "bulletMaterial": Object {
        "examples": Array [
          Object {
            "acsModelName": "COPA MUNDIAL",
            "b2bCopy": "Hurricanes Training Jersey as worn by the players when training with dynamic retro design inspired by the original Super Jersey.Captured with shock blue and black colors on climalite fabric with moisture wicking properties.",
            "b2bCopyShort": "Hurricanes Training Jersey as worn by the players when training with dynamic retro design inspired by the orginal Super jersey.",
            "b2cCopy": "De populairste voetbalschoen ooit, en terecht. De adidas Copa Mundial heeft een bovenwerk van rijk leer en een loopzool met dubbele dichtheid die al decennia lang de toon zet voor comfort en prestaties op het veld.",
            "b2cCopyShort": "Entraîne-toi comme les pros avec ce haut de l'équipe de Belgique de football.",
            "b2cModelName": "COPA MUNDIAL",
            "bulletFit": "bulletFit",
            "bulletMaterial": "Maille double 100 % polyester.",
            "bulletMeasurements": "bullet measurements",
            "bulletWeight": "bullet weight",
            "bullets1": "materiál Climalite odvádí pot a udrží tě vsuchu za všech okolností",
            "bullets2": "Ribbet V-hals",
            "bullets3": "Nackenband",
            "bullets4": "Επένδυση προστασίας από τον άνεμο στο μπροστινό μέρος. Μανσέτες με ανοίγματα για τους αντίχειρες",
            "bullets5": "Woven Belgium crest on chest",
            "bullets6": "3 bandas en las mangas; logotipo de adidas serigrafiado a la derecha del pecho",
            "bullets7": "Three stripes",
            "locale": "en-GB",
            "materialComposition": "100% Organic Cotton",
            "materialConstruction": "Single jersey",
          },
        ],
        "type": "string",
      },
      "bulletMeasurements": Object {
        "type": "string",
      },
      "bulletWeight": Object {
        "type": "string",
      },
      "bullets1": Object {
        "type": "string",
      },
      "bullets2": Object {
        "type": "string",
      },
      "bullets3": Object {
        "type": "string",
      },
      "bullets4": Object {
        "type": "string",
      },
      "bullets5": Object {
        "type": "string",
      },
      "bullets6": Object {
        "type": "string",
      },
      "bullets7": Object {
        "type": "string",
      },
      "locale": Object {
        "type": "string",
      },
      "materialComposition": Object {
        "type": "string",
      },
      "materialConstruction": Object {
        "type": "string",
      },
    },
    "title": "Catalogues",
    "type": "object",
  },
  Object {
    "$id": "http://supermodel.io/adidas/product/Color",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "description": "Color descriptions are language-dependent",
    "examples": Array [
      Object {
        "color1": Object {
          "code": "A0QM",
          "description": "core black",
        },
        "color2": Object {
          "code": "A98B",
          "description": "utility black",
        },
        "color3": Object {
          "code": "A0QM",
          "description": "core black",
        },
        "color4": Object {
          "code": "01F7",
          "description": "ftwr white",
        },
      },
    ],
    "properties": Object {
      "color1": Object {
        "$ref": "utils/CodeDescription",
      },
      "color2": Object {
        "$ref": "utils/CodeDescription",
      },
      "color3": Object {
        "$ref": "utils/CodeDescription",
      },
      "color4": Object {
        "$ref": "utils/CodeDescription",
      },
    },
    "title": "Color",
    "type": "object",
  },
  Object {
    "$id": "http://supermodel.io/adidas/product/SKU",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "examples": Array [
      Object {
        "EAN": "4058026055128",
        "UPC": "885591427616",
        "conversionIDs": Array [
          Object {
            "saleSizeAbbr": "27",
            "saleSizeConvID": "D",
            "saleSizeScale": "8C",
          },
        ],
        "srcSizeCD3": "380",
        "srcSizeScale": "T4",
      },
    ],
    "properties": Object {
      "EAN": Object {
        "type": "string",
      },
      "UPC": Object {
        "type": "string",
      },
      "conversionIDs": Object {
        "items": Array [
          Object {
            "properties": Object {
              "localSize": Object {
                "type": "string",
              },
              "saleSizeAbbr": Object {
                "type": "string",
              },
              "saleSizeConvID": Object {
                "type": "string",
              },
              "saleSizeDesc": Object {
                "type": "string",
              },
              "saleSizeScale": Object {
                "type": "string",
              },
            },
            "type": "object",
          },
        ],
        "type": "array",
      },
      "srcSizeCD3": Object {
        "type": "string",
      },
      "srcSizeDescription": Object {
        "type": "string",
      },
      "srcSizeScale": Object {
        "type": "string",
      },
    },
    "title": "SKU",
    "type": "object",
  },
  Object {
    "$id": "http://supermodel.io/adidas/product/utils/CodeDescription",
    "$schema": "http://json-schema.org/draft-07/schema#",
    "description": "Helper data structure to hold any code - description pair",
    "properties": Object {
      "code": Object {
        "type": "string",
      },
      "description": Object {
        "type": "string",
      },
    },
    "title": "Code Description Tuple",
    "type": "object",
  },
]
`;
